File: PythonProblems/Sorting/MergeSort.py
'''
Implementation of merge sort
'''

def merge_sort(arr):
    """
    Args:
     arr(list_int32)
    Returns:
     list_int32
    """
    # Write your code here.
    print("arr:",arr)
    if(len(arr)<=1):
        return arr
    #Get midpoint
    mid = int(len(arr)/2)
    #Recursive sort left, sort right
    leftarr = merge_sort(arr[:mid])
    rightarr = merge_sort(arr[mid:])
    print("Sorted letf:",leftarr)
    print("Sorted right:",rightarr)
    #combine
    i,j,k = 0,0,0
    while(i<len(leftarr) and j<len(rightarr)):
        if(leftarr[i]<rightarr[j]):
            arr[k]=leftarr[i]
            i=i+1
        elif(leftarr[i]>rightarr[j]):
            arr[k]=rightarr[j]
            j=j+1
        elif(leftarr[i]==rightarr[j]):
            arr[k]=leftarr[i]
            k=k+1
            arr[k]=rightarr[j]
            i,j = i+1, j+1
        k = k+1
    while(i<len(leftarr)):
        arr[k]=leftarr[i]
        i=i+1
        k=k+1
    while(j<len(rightarr)):
        arr[k]=rightarr[j]
        j=j+1
        k=k+1
    print("Sorted array:",arr)
    return arr

#arr = [5, 8, 3, 9, 4, 1, 7]
#arr = [-1,-2,-3,-4,-5,-6]
#arr = [-1,-5,-4,-3,-2,-10]
arr = [-1,-2,-3,-4]
#print("Initial:",arr)
sortarr = merge_sort(arr)
print("Sorted arr:",sortarr)

File: CPP/Sorting/MergeSort.cpp
class Solution {
public:
    void merge(vector<int> &arr, int left, int mid, int right, vector<int> &tempArr)
    {
        int start1 = left;
        int start2 = mid+1;
        int end1 = mid-left+1;
        int end2 = right-mid;
        for(int i=0;i<end1;i++) tempArr[start1+i] = arr[start1+i];
        for(int i=0;i<end2;i++) tempArr[start2+i] = arr[start2+i];

        //Merge sub-arrays
        int i=0,j=0,k=left;
        while(i<end1 && j<end2)
        {
            if(tempArr[start1+i]<tempArr[start2+j])
            {
                arr[k]=tempArr[start1+i];
                i++;
            }
            else
            {
                arr[k]=tempArr[start2+j];
                j++;
            }
            k++;
        }
        while(i<end1)
        {
            arr[k]=tempArr[start1+i];
            i++,k++;
        }
        while(j<end2)
        {
            arr[k]=tempArr[start2+j];
            j++,k++;
        }
    }
    
    void mergeSort(vector<int> &arr, int left, int right, vector<int> &tempArr)
    {
        if(left>=right) return;
        //Get midpoint
        int mid = (left+right)/2;
        mergeSort(arr,left,mid,tempArr);
        mergeSort(arr,mid+1,right,tempArr);
        //Combine
        merge(arr,left,mid,right,tempArr);
    }

    vector<int> sortArray(vector<int>& nums) {
        vector<int> temp1Array(nums.size());
        mergeSort(nums,0,nums.size()-1, temp1Array);    
        return nums;
    }
};